nifi-extension-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/GetS3ObjectMetadata.java

```diff
@@ -0,0 +1,192 @@
+            .addValidator(StandardValidators.NON_EMPTY_EL_VALIDATOR)
+            .defaultValue("s3.object.metadata")
+            .dependsOn(METADATA_TARGET, TARGET_ATTRIBUTE)
+            .required(true)
+            .build();
+
+    public static final ObjectMapper MAPPER = new ObjectMapper();
+
+    public static final List<PropertyDescriptor> properties = List.of(
+            MODE,
+            METADATA_TARGET,
+            METADATA_ATTRIBUTE,
+            BUCKET_WITH_DEFAULT_VALUE,
+            KEY,
+            AWS_CREDENTIALS_PROVIDER_SERVICE,
+            S3_REGION,
+            TIMEOUT,
+            FULL_CONTROL_USER_LIST,
+            READ_USER_LIST,
+            READ_ACL_LIST,
+            OWNER,
+            SSL_CONTEXT_SERVICE,
+            ENDPOINT_OVERRIDE,
+            SIGNER_OVERRIDE,
+            S3_CUSTOM_SIGNER_CLASS_NAME,
+            S3_CUSTOM_SIGNER_MODULE_LOCATION,
+            PROXY_CONFIGURATION_SERVICE);
+
+    public static Relationship REL_FOUND = new Relationship.Builder()
+            .name("found")
+            .description("An object was found in the bucket at the supplied key")
+            .build();
+
+    public static Relationship REL_NOT_FOUND = new Relationship.Builder()
+            .name("not found")
+            .description("No object was found in the bucket the supplied key")
+            .build();
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return new HashSet<>(List.of(REL_FOUND, REL_NOT_FOUND, REL_FAILURE));
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        boolean isRouter = context.getProperty(MODE).getValue().equals(MODE_ROUTER.getValue());
+
+        final AmazonS3Client s3;
+        try {
+            s3 = getS3Client(context, flowFile.getAttributes());
+        } catch (Exception e) {
+            getLogger().error("Failed to initialize S3 client", e);
+            flowFile = session.penalize(flowFile);
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+
+        final String bucket = context.getProperty(BUCKET_WITH_DEFAULT_VALUE).evaluateAttributeExpressions(flowFile).getValue();
+        final String key = context.getProperty(KEY).evaluateAttributeExpressions(flowFile).getValue();
+
+        try {
+            Relationship route;
+
+            try {
+                ObjectMetadata metadata = s3.getObjectMetadata(bucket, key);
+                Map<String, Object> combinedMetadata = new HashMap<>(metadata.getRawMetadata());
+                combinedMetadata.putAll(metadata.getUserMetadata());
+
+                String metadataJson = MAPPER.writeValueAsString(combinedMetadata);
```

Comment:
Rather than turning the entire metadata map into a JSON string, what about setting each metadata key as the attribute key, and then just turning the attribute value into a String. We could consider an Attributes JSON strategy as well, but mapping the metadata directly to attributes has some benefits.

Reply:
@MikeThomsen I totally agree with @exceptionfactory that it is more useful to store the metadata in separate attributes.

Here is a sample S3 metadata JSON that I found.
```
{
    "DocumentId": "S3 document ID, the S3 path to doc",
    "Attributes": {
        "_category": "document category",
        "_created_at": "ISO 8601 encoded string",
        "_last_updated_at": "ISO 8601 encoded string",
        "_source_uri": "document URI",
        "_version": "file version",
        "_view_count": number of times document has been viewed,
        "custom attribute key": "custom attribute value",
        additional custom attributes
    },
    "AccessControlList": [
         {
             "Name": "user name",
             "Type": "GROUP | USER",
             "Access": "ALLOW | DENY"
         }
    ],
    "Title": "document title",
    "ContentType": "For example HTML | PDF. For supported content types, see Types of documents."
}
```

Reply:
I think each of the fields under Attributes would be useful to store as separate attributes.
I mean Attributes._category (or just _category if you prefer and if you do not think the subfield names will collide with top level names DocumentId, Title, etc) for the attribute name and so on.
I would store the AccessControlList as a single attribute with its JSON array as the attribute value because I predict that info will be used less often.